### PR TITLE
feat: webln integration for blocktank

### PR DESCRIPTION
**Pull request by pavanjoshi914**
webln integration for lightning payments.

https://github.com/synonymdev/blocktank-ui/assets/55848322/2d3893a8-ab34-44d0-8cd8-10004c97e2ec

### CODE REVIEW

**Comment by BitcoinErrorLog** [comment]
Is this also working properly when Alby is disabled?

**Comment by pavanjoshi914** [comment]
> Is this also working properly when Alby is disabled?

unfortunately, there is no way to detect if the wallet is disabled. alby and other lightning providers do inject providers even if it's disabled. it's just that providers won't work

**Comment by pavanjoshi914** [comment]
+ are there any docs regarding how to test blocktank widget in the dev environment? do I have to spend my own sats to reach to claim channel page?

**Comment by rbndg** [comment]
We currently test internally with regtest, so we would have to merge and test it ourselves.

**Comment by rbndg** [comment]
After a review, we can merge this and we can test it internally

**Comment by pavanjoshi914** [comment]
any updates on this pr? is there anything i can help with?

**Comment by flaviomoceri** [comment]
Hi @pavanjoshi914,
the widget has been updated, now you can test in regtest simply by setting the environment variable `API_URL="https://api.stag.blocktank.to/blocktank/api/v2/"`. If you don't know how to make payment in regtest you can use bitkit or provide me with the LN invoice and I will pay it for you.

I checked the code, the payment works, these are the errors:
1. Add space between "Claim Now" button and "Copy claim url" (you can use `<div className={'claim-channel-button-spacer'} />`)
2. Replace `onClick={() => ClaimWithWebln}` with `onClick={ClaimWithWebln}` otherwise the function is not called
3. When I try to claim the channel via Alby it doesn't seem to work (please take a look)

If you need help or have any other questions, please let me know.

I am on mainnet

https://github.com/synonymdev/blocktank-ui/assets/63161412/97e31bbd-a709-4360-8e5c-bdcdd1da154b

**Comment by pavanjoshi914** [comment]
i am not sure how i can access the claim page. is there any testnet setup you people provide?

**Comment by flaviomoceri** [comment]
You can test in mainnet, currently regtest is only for internal testing

But if you use this [PR](https://github.com/synonymdev/blocktank-ui/pull/78/) and set the environment variable `REACT_APP_API_URL="https://api.stag.blocktank.to/blocktank/api/v2/"`, you should be able to create an order in regtest and if you send me LN invoice regtest I will pay it for you, so you will be able to see the "Claim Channel" page

**Comment by pavanjoshi914** [comment]
thanks claim page might not have been working because of typo. corrected the function call should work now

**Comment by flaviomoceri** [comment]
Great, please try claiming a channel by making a video to make sure everything works

You can test on regtest by following these instructions https://github.com/synonymdev/blocktank-ui/pull/73#issuecomment-1941492765, otherwise if you stay on mainnet, if you send me the LN invoice I will pay for it for you, so you can make sure that everything works

**Comment by pavanjoshi914** [comment]
if you want me to try mainnet here's my lightning address pavanj@getalby.com

**Comment by flaviomoceri** [comment]
Done, sent

**Comment by pavanjoshi914** [comment]
@flaviomoceri  which node you use in alby. few of the connector implementations doesn't support connectPeer() which is required while claiming channel. https://github.com/getAlby/lightning-browser-extension/pull/3041

i created pr in wallet to show proper error when method is not supported. 

image


not it shows proper error message

**Comment by flaviomoceri** [comment]
I currently don't have any node connected to my alby, please use your node to test, a video is more than enough

**Comment by pavanjoshi914** [comment]
https://github.com/synonymdev/blocktank-ui/pull/73#issuecomment-1931572115 yup i am just clarifying the video mentioned here.

**Comment by pavanjoshi914** [comment]
image

do you know why server is responding with status code 400 ? 
here's the lnurl of created channel
lnurl1dp68gurn8ghj7cnvda3kkarpde4juumedehku7td9e6x7tmpwp5j7a3j9akz7v34xd3rqdmrxuknzve3x5kngdekxckkyvnxxuknverpxasnyerzx93kgcstylxh6

**Comment by flaviomoceri** [comment]
When do you get this error and what kind of problem are you having while trying to claim with alby?

**Comment by pavanjoshi914** [comment]
while trying to claim the channel. 
lnurl decoder shows this
image

**Comment by flaviomoceri** [comment]
https://blocktank.synonym.to/api/v2/channels/253b07c7-1315-4766-b2f7-6da7a2db1cdb

It says the payment status is canceled, so the transaction has been canceled and you should have received your money back, please try again with a new channel and let me know if you get the same error

**Comment by SeverinAlexB** [comment]
> do you know why the server is responding with status code 400? here's the lnurl of created channel lnurl1dp68gurn8ghj7cnvda3kkarpde4juumedehku7td9e6x7tmpwp5j7a3j9akz7v34xd3rqdmrxuknzve3x5kngdekxckkyvnxxuknverpxasnyerzx93kgcstylxh6

Hi @pavanjoshi914. I tested the LNURL endpoint on mainnet manually again and it works just fine for me. If you can reproduce this error again, can you show me the exact network request/response that shows the 400 error? 400 hints to a request validation issue. Maybe Alby is using LNURL slightly different than we expect it.

Where can you find this request/response?

Developer Console -> Network -> Select the correct request -> Show me Payload, Response, and Headers/General section.
CleanShot 2024-02-15 at 08 38 22@2x

**Comment by pavanjoshi914** [comment]
i figured out alby lnurl channel implementation. found that they were missing a private field to add in lnurl callback will be fixed with merging of this pr https://github.com/getAlby/lightning-browser-extension/pull/3046

**Comment by pavanjoshi914** [comment]
lnurl channel fixes are merged! and will be present in the next release.  please do test it. latest changes are present in alby nightly builds https://github.com/getAlby/lightning-browser-extension#try-out-the-most-recent-version-of-alby-nightly-releases

**Comment by flaviomoceri** [comment]
Great @pavanjoshi914, so have you tried claiming a channel via alby and got no errors?

**Comment by flaviomoceri** [comment]
any updates? We tried to test by connecting the node on alby but without success

**Comment by MiguelMedeiros** [comment]
Hey @pavanjoshi914, thanks for your PR! 🥳 

I've connected my node to my Alby extension (with nightly build) and tried to claim the channel with it. 
But I got some errors! 😢 

1) First try:

Nothing happend and I remebered that I needed to use the nightly build). 😄 

https://github.com/synonymdev/blocktank-ui/assets/5798170/1d65fce1-b9de-45a1-90ac-89f2b2265f82

2) Second try:

Now with the nightly build I've tried again. Got this error:

**Error: Not yet supported with the currently used account.**

https://github.com/synonymdev/blocktank-ui/assets/5798170/229d6836-95df-44b5-9e45-9d67766eedb8

3) Third try:

Tried again with the option "private channel", and got the same errror:

**Error: Not yet supported with the currently used account.**

https://github.com/synonymdev/blocktank-ui/assets/5798170/1879ffe7-6efa-4e0c-a3d0-40a3fc199afe


Do I need some special permissions to integrate that with my Alby account?

**Comment by pavanjoshi914** [comment]
hey @MiguelMedeiros every connector needs connectPeer function to make lnurl-withdraw work. unfortunately all alby connectors doesn't have that method implemented due to some issues. currently it's supported by only two connectors commando and lnd. if user has such nodes connected via alby it will work only for that.

maybe we just use webln only for payments and not lnurl channel opening for now? wdyt?

**Comment by pavanjoshi914** [comment]
Thanks resolved the conflicts

File: src/pages/public/claim/index.tsx [hunk 1]
```diff
@@ -84,6 +98,12 @@ function ClaimPage(): JSX.Element {
 					</div>
 					<p className={'claim-channel-address'}>{clipCenter(lnurl.toLocaleLowerCase(), 42)}</p>
 					<div>
+						{typeof window.webln !== 'undefined' && window.webln.lnurl && (
+							<>
+								<ActionButton onClick={ClaimWithWebln}>Claim Now</ActionButton>
+								<div className={'claim-channel-button-spacer'} />
+							</>
+						)}
```
Comment (MiguelMedeiros) [inline src/pages/public/claim/index.tsx:106]:
```suggestion
```

@pavanjoshi914 I think we should just remove this part, right?

File: src/pages/public/claim/index.tsx [hunk 2]
```diff
@@ -50,6 +50,18 @@ function ClaimPage(): JSX.Element {
...
+		if (typeof window.webln !== 'undefined' && window.webln.lnurl) {
+			try {
+				await window.webln.enable();
+
+				await window.webln.lnurl(lnurl);
+			} catch (e) {
+				alert(`An error occurred during the payment: ${e}`);
+			}
+		}
+	};
```
Comment (MiguelMedeiros) [inline src/pages/public/claim/index.tsx:63]:
```suggestion
```

Also remove this function.

**Comment by pavanjoshi914** [comment]
reverted webln changes from claim page

**Review by MiguelMedeiros — APPROVED** [review]
ACK. 👍

**Pull request merged**